x-pack/plugins/ui_actions_enhanced/server/telemetry/dynamic_action_factories_collector.ts

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { DynamicActionsState } from '../../common';
+import { ActionFactory } from '../types';
+import { getMetricKey } from './get_metric_key';
+
+export const dynamicActionFactoriesCollector = (
+  getActionFactory: (id: string) => undefined | ActionFactory,
+  state: DynamicActionsState
+): Record<string, any> => {
+  const stats: Record<string, any> = {};
+
+  for (const event of state.events) {
+    const factory = getActionFactory(event.action.factoryId);
+
+    if (factory) {
+      let factoryStats: Record<string, any> = {};
```

Comment:
by passing a new stats object to each `factory.telemetry` function you prevent factories from logging/counting same events, which was something we wanted to support.

Reply:
is there specific reasons not to pass in the original stats object ?

Reply:
That is something I wanted to discuss. I currently don't see how it could work otherwise. The actions report their telemetry, they don't know that they are used on a dashboard in an embeddable on some specific panel. Let's say at the end we want to construct keys like this

```
dashboard.embeddable.dynamicActions.count
dashboard.<dashbaordId>.embeddable.dynamicActions.count
```

UI actions don't know about the key prefixes `dashboard`, `embeddable`, `<dashboardId>`.

Reply:
if you just pass the stats thru (skip the local vars), everything this pr adds will still work, + it will leave additional options open for the future. or am I missing something ?

Reply:
I've implemented it as we discussed over Zoom.